Style guide on optional arguments for parents of return values (#2084)
oscar-system/Oscar.jl: docs/src/DeveloperDocumentation/styleguide.md

However, as always, rules sometimes should be broken.

## Optional arguments for parents of return values

> Several objects in OSCAR have `parent`s, e.g. polynomials, group elements, ... 

Reviewer: You could also link to the relevant section of the AbstractAlgebra documentation here?

Reviewer 2: No clue how to do that.

Reviewer: Well...
```suggestion
Several objects in OSCAR have `parent`s, e.g. polynomials, group elements, and so on.
See [here](https://docs.oscar-system.org/stable/Nemo/developer/parents/) for more information.
```

Whenever a function creates such objects from an input which does not involve 

> the output's parent, the user should have the possibility to pass on this 

Reviewer: ```suggestion
the output's parent, the caller should have the possibility to specify this 
```

> parent as an optional argument. 

Reviewer: So here it specifically talks about an "optional" argument. But then in the example it shows *two* methods for specifying the caller, only one involving an optional argument. Nothing seems to resolve this?

> Let's see an example. Say, you want to implement the characteristic 

Reviewer: ```suggestion
As an example, suppose you want to add a function computing the characteristic 
```

> polynomial of a matrix. You could do it as follows:

Reviewer: ```suggestion
polynomial of a matrix (OSCAR of course already has this, but we will ignore this
for the moment for the sake of this example). You could implement it as follows:
```

```julia
  function characteristic_polynomial(A::MatrixElem)
    kk = base_ring(A)

>     P, x = kk["x"]

Reviewer: ```suggestion
    P, x = kk[:x]
```

Reviewer 3: both are terrible - and should be replaced by `polynomial_ring(kk, cached = false)`. Within the system, `kk[:x]` and variants should never be used as they create global cached objects that clutter up the memory.

    AP = change_base_ring(P, A)
    return det(AP - x*one(AP))
  end
```

> You can see that the polynomial ring `P`, i.e. the parent of the output, 

Reviewer: ```suggestion
The polynomial ring `P`, i.e. the parent of the output, 
```

is newly created in the body of the function. In particular, calling this 
function two times on two different matrices `A` and `B` might produce 
incompatible polynomials `p = det(A - x*one(A))` and `q = det(B - x*one(B))` 
with different parents. Then `p + q` will produce an error. 

> To solve this, we should have implemented the function differently:

Reviewer: ```suggestion
To overcome this, the function should allow specifying the parent
of the return value:
```

```julia
  function characteristic_polynomial(
      A::MatrixElem;

>       ring::AbstractAlgebra.Ring=(base_ring(A)["t"])[1]

Reviewer 2: ```suggestion
      parent::AbstractAlgebra.Ring=(base_ring(A)["t"])[1]
```

>       parent::AbstractAlgebra.Ring=(base_ring(A)["t"])[1]

Reviewer: ```suggestion
      parent::AbstractAlgebra.Ring=(base_ring(A)[:t])[1]
```

    )

>     AP = change_base_ring(ring, A)

Reviewer 2: ```suggestion
    AP = change_base_ring(parent, A)
```

>     x = first(gens(ring))

Reviewer 2: ```suggestion
    x = first(gens(parent))
```

>     x = first(gens(parent))

Reviewer: Ironically, in this example it seems to me that specifying the intended generator `x` would be even nicer... ;-)

    return det(AP - x*one(AP))
  end
  function characteristic_polynomial(

>       P::AbstractAlgebra.Ring,

Reviewer 2: ```suggestion
      parent::AbstractAlgebra.Ring,
```

      A::MatrixElem
    )
    coefficient_ring(P) === base_ring(A) || error("coefficient rings incompatible")

>     return characteristic_polynomial(A, ring=P)

Reviewer 2: ```suggestion
    return characteristic_polynomial(A, parent=P)
```

  end
```
This allows for two different entry points for the ring `P` of the output: 
Once as the first argument of a method of `characteristic_polynomial` with 

> an extended signature, and second as an optional keyword argument for `ring` in 

Reviewer 2: ```suggestion
an extended signature, and second as an optional keyword argument for `parent` in 
```

the original method. This should be the general rule for such implementations 

> within OSCAR. 

Reviewer: I still disagree with this: I don't want two ways to do this in general. There should be one way, and that should be implemented consistently (instead of some code implementing one, some the other, some both)

Reviewer 2: Agreed. It is better to be consistent and have only one variant.  I have no preference here.

Reviewer 2: Rethinking I would prefer 
`f(parent, input)`
`f(input)`
over 
`f(input; parent)`
`f(input)`
This matches our parent first philosophy for e.g. `ideal` and soon `factor`.
@fingolfin @thofma @fieker  any opinion?

Reviewer: @fieker just pointed out an issue with this approach: some functions like `map_coefficients` have another "canonical" first argument, namely a function to apply to all arguments. Here is one declaration:
```julia
map_coefficients(f, p::PolyRingElem{<: RingElement};
  cached::Bool=true, parent::PolyRing)
```

  function characteristic_polynomial(
      A::MatrixElem;
      parent::AbstractAlgebra.Ring=polynomial_ring(base_ring(A), :t)[1]
    )
    AP = change_base_ring(parent, A)
    x = first(gens(ring))
    return det(AP - x*one(AP))
  end

  # Optional second signature to also allow for the specification of the 
  # output's parent as the first argument:
  function characteristic_polynomial(
      P::PolyRing,
      A::MatrixElem
    )
    coefficient_ring(P) === base_ring(A) || error("coefficient rings incompatible")
    return characteristic_polynomial(A, parent=P)
  end
```
In fact this now allows for two different entry points for the parent ring `P` 
of the output: First as the required `parent` keyword argument and second 
as the first argument of a method of `characteristic_polynomial` with 
an extended signature. Note that within the scope of the first method's body 
the OSCAR function `parent` is necessarily overwritten by the name of the 
keyword argument. Hence to call the actual parent of any other object, you 
must then use `Oscar.parent`. E.g. to get the `MatrixSpace` of the 
matrix `A`, write `Oscar.parent(A)`.

> 

Reviewer: @fieker mentioned a third option that is sometimes used: passing in a ring element (or at least a polynomial) and then the output has the same parent resp. is evaluated at that input ring element. E.g.:
```
@doc raw"""
    cyclotomic(n::Int, x::ZZPolyRingElem)

Return the $n$th cyclotomic polynomial, defined as
$$\Phi_n(x) = \prod_{\omega} (x-\omega),$$ where $\omega$ runs over all the
$n$th primitive roots of unity.
"""
function cyclotomic(n::Int, x::ZZPolyRingElem)
```
